### ISSUE TITLE
LSTM Model Fails to Converge

### ISSUE THREAD

**Issue by kumarutkarsh1248** [mlpack/mlpack#3953]
I've been trying to train a simple LSTM model in mlpack for a many-to-one prediction task — basically, it takes 50 points from a sine function and predicts the next one.

But the model just doesn't seem to converge. Even after 50 epochs, the loss stays around 0.333 and doesn’t go down much. The training setup is pretty standard, so I expected it to perform better.

To cross-check, I used the same architecture, optimizer, learning rate, batch size, and data in PyTorch — and that model converged really fast, with the loss dropping close to zero within 50 epochs.

So now I’m wondering if there's a bug in mlpack's LSTM implementation or how it's being trained

[link to the colab notebook](https://colab.research.google.com/drive/19reUxVtZemtCI8nMhFkUuBZ4ZCi-SiZC?usp=sharing)

mlpack code:
```cpp

#include <mlpack.hpp>

using namespace ens;
using namespace mlpack;
using namespace std;

void GenerateSineSeq(
    arma::cube &data,
    arma::cube &labels,
    int samples = 1000,
    int time_step = 50)
{
  // Generate 1000 points from 0 to 100
  arma::vec x = arma::linspace(0, 100, samples + time_step);

  data = arma::cube(1, samples, time_step);
  labels = arma::cube(1, samples, time_step);

  for (int i = 0; i < samples; i++)
  {
    data.tube(0, i) = x.subvec(i, i + time_step - 1);
    labels(0, i, time_step - 1) = x(i + time_step);
  }
  data = arma::sin(data);
  labels = arma::sin(labels);

  return;
}

int main()
{
  arma::cube input, labels;
  GenerateSineSeq(input, labels);

  RNN<MeanSquaredError> model(50, true /* only one response per sequence */);
  model.Add<Identity>();
  model.Add<LSTM>(32);
  model.Add<Linear>(1);

  // Set parameters for the Adam optimizer.
  int EPOCHS = 50;
  ens::Adam optimizer(0.01, 1, 0.9,
                      0.999,
                      1e-8,
                      input.n_cols * EPOCHS,
                      1e-8,
                      true);

  model.Train(input,
              labels,
              optimizer,
              ens::PrintLoss(),
              ens::ProgressBar(),
              ens::EarlyStopAtMinLoss());

  return 0;
}

```

**Comment by nzachary** [comment]
Hi @kumarutkarsh1248 

I believe the error is in your code. Even though a RNN will accept those labels without an error, an RNN in single mode actually expects labels with only 1 slice. In this example, your model is actually training on whatever is in the first slice (probably a 0) rather than the next point in the sine wave.

Assuming there isn't a bug in mlpack's LSTM or RNN implementation, replacing `labels = arma::cube(1, samples, time_step);` and `labels(0, i, time_step - 1) = x(i + time_step);` with `labels = arma::cube(1, samples, 1);` and `labels(0, i, 0) = x(i + time_step);` should allow your model to converge.

**Comment by kumarutkarsh1248** [comment]
Actually I was following this [link](https://github.com/mlpack/models/issues/50#issuecomment-803435373)

What you're suggesting also seems correct, as shown in this example [link](https://www.mlpack.org/doc/tutorials/ann.html#:~:text=size_t%20numColumns%20%3D%20y.n_elem%20/%20rho%3B%0A%20%20data%20%3D%20arma%3A%3Acube(1%2C%20numColumns%2C%20rho)%3B%0A%20%20labels%20%3D%20arma%3A%3Acube(1%2C%20numColumns%2C%201)%3B)


However, I’ve tried both approaches, and neither of them seems to converge.
If your model is converging, please share your code here.

**Comment by eddelbuettel** [comment]
@kumarutkarsh1248  not sure what you are trying to do with those 'links' but it isn't working:
was the '-So%2C%20the%20shape)' supposed to be part of a link? 

You can 'Preview' a post to test your links.  The markup style is markdown, and it is [documented](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).

**Comment by nzachary** [comment]
@kumarutkarsh1248 The discussion in the first link is from Mar 11, 2021 and the newest mlpack release at that time was 3.4.2. The code in that discussion is using a version that is whole major version behind the current version, 4.6.2. While it may have been correct at the time, the library has changed a lot since then.

The second link links to the documentation so it should be correct and up to date with the current API. 

Although I had to lower the learning rate, this model converges and the results actually [look better than the results in your colab](https://files.catbox.moe/vv0tta.png)
```c++
#include <mlpack.hpp>

using namespace ens;
using namespace mlpack;
using namespace std;

void GenerateSineSeq(
    arma::cube &data,
    arma::cube &labels,
    int samples = 1000,
    int time_step = 50)
{
  // Generate 1000 points from 0 to 100
  arma::vec x = arma::linspace(0, 100, samples + time_step);

  data = arma::cube(1, samples, time_step);
  labels = arma::cube(1, samples, 1);

  for (int i = 0; i < samples; i++)
  {
    data.tube(0, i) = x.subvec(i, i + time_step - 1);
    labels(0, i, 0) = x(i + time_step);
  }
  data = arma::sin(data);
  labels = arma::sin(labels);

  return;
}

int main()
{
  arma::cube input, labels;
  GenerateSineSeq(input, labels);

  RNN<MeanSquaredError> model(50, true /* only one response per sequence */);
  model.Add<Identity>();
  model.Add<LSTM>(32);
  model.Add<Linear>(1);

  // Set parameters for the Adam optimizer.
  int EPOCHS = 50;
  ens::Adam optimizer(1e-4, 1, 0.9,
                      0.999,
                      1e-8,
                      input.n_cols * EPOCHS,
                      1e-8,
                      true);

  model.Train(input,
              labels,
              optimizer,
              ens::PrintLoss(),
              ens::ProgressBar(),
              ens::EarlyStopAtMinLoss());

  arma::cube predictors, out, next;
  out = arma::cube(1, 1, 200);
  // Prepare data
  predictors = arma::cube(1, 1, 1000);
  predictors.tube(0, 0) = arma::linspace(0, 100, 1000);
  predictors = arma::sin(predictors);
  ((arma::mat)predictors.tube(0, 0)).save("groundtruth.csv", arma::csv_ascii);
  predictors = predictors.slices(0, 49);

  for (int i = 0; i < 200; i++)
  {
    model.Predict(predictors, next);
    // Shift predictors down and append next
    for (int j = 0; j < 49; j++)
    {
      predictors.slice(j) = predictors.slice(j + 1);
    }
    predictors.slice(49) = next.slice(0);
    // Write the predicted next value to out
    out.slice(i) = next.slice(0);
  }
  ((arma::mat)out.tube(0, 0)).save("out.csv", arma::csv_ascii);

  return 0;
}
```

**Comment by kumarutkarsh1248** [comment]
@nzachary You're right — the results are much better now. Thanks for your help! 
Could you please share the final loss value your model reaches at the end of training

**Comment by nzachary** [comment]
Glad I could help!
The loss at epoch 50 was 0.00014115.

**Comment by kumarutkarsh1248** [comment]
I ran your exact same code on my system four times, and in each of these runs, I got identical results.
The training stopped early at epoch 19 with a final loss of 11.6273 every time.
This early stopping appears to be due to no improvement in loss across epochs.

It seems that mlpack uses a fixed random seed internally, which causes the same initialization and hence the same results in each seprate training.

To test this, I manually set different seeds using: `arma::arma_rng::set_seed(171);`

And the results are as follows:
| seed | epochs | loss |
|----------|----------|----------|
| 22    | 16     | 20.3159     |
| 2033    | 18    | 19.4208     |
| 17    | 16    | 23.2017     |
| 171    | 16    | 11.9892     |


![Image](https://github.com/user-attachments/assets/05e79610-75f7-4b63-a3ed-804c5a28e577)

This is the graph I obtained when the random seed was set to 171.
The model definitely seems to learn, and the results are much better now. However, I'm trying all these variations to achieve even better results similar to yours.

Do you have any suggestions?

**Comment by nzachary** [comment]
Ideally you should be getting similar results to mine without needing any changes. Can you please provide the following information about your environment? : mlpack version, armadillo version, ensmallen version, compiler, compiler flags (if applicable), operating system, cpu, and anything else you think might be important.

**Comment by kumarutkarsh1248** [comment]
mlpack version: 4.4.1
Armadillo version: 10.8.2
ensmallen version: 2.18.2
compiler version: g++ 11.4.0
OS: Ubuntu 22.04.5
CPU: x86_64, AMD Ryzen 3 3250U

**Comment by nzachary** [comment]
> mlpack version: 4.4.1

4.4.1 doesn't exist; did you mean 4.4.0? Try updating to the latest version. It wasn't listed in the release notes but ~~4.5.1~~ 4.6.0 revamped the LSTM layer and fixed some issues with BPTT.

Edit: I was able to reproduce the issue on <= 4.5.1 but it works fine on >= 4.6.0.

**Comment by kumarutkarsh1248** [comment]
My accuracy is now matching yours. Thanks for your help!

**Comment by rcurtin** [comment]
Awesome, seems like this is fixed then so I'll go ahead and close the issue. :+1:
